src/v/io/tests/scheduler_test.cc

```diff
@@ -0,0 +1,194 @@
+    scheduler_tester(
+      io::scheduler* scheduler,
+      io::persistence* storage,
+      std::filesystem::path path,
+      size_t file_size,
+      size_t num_io_ops)
+      : scheduler(scheduler)
+      , queue(
+          storage,
+          std::move(path),
+          [this](auto& page) noexcept { generator.handle_complete(page); })
+      , generator(
+          file_size,
+          num_io_ops,
+          [this, scheduler](auto& page) {
+              scheduler->submit_read(&queue, &page);
+          },
+          [this, scheduler](auto& page) {
+              scheduler->submit_write(&queue, &page);
+          })
+      , faulter{
+          storage,
+          io::testing_details::scheduler_queue_accessor::get_queue(&queue),
+          10,
+          100 // min,max delay between faults
+        } {}
+
+    seastar::future<> run() {
+        scheduler->add_queue(&queue);
+        faulter.start();
+        return generator.run();
+    }
+
+    void stop() {
+        faulter.stop().get();
+        io::scheduler::remove_queue(&queue).get();
+    }
+
+    io::scheduler* scheduler;
+    io::scheduler::queue queue;
+    io_queue_workload_generator generator;
+    io_queue_fault_injector faulter;
+};
+
+/*
+ * This test uses a load generator to run reads and writes against multiple I/O
+ * queues while limiting the maximum number of open file handles. Faults are
+ * injected during execution.
+ */
+TEST_P(SchedulerTest, WriteRead) {
+    // paths for test files. one per queue.
+    std::vector<std::filesystem::path> paths;
+    for (auto i : boost::irange(num_queues())) {
+        paths.emplace_back(fmt::format("foo.{}", i));
+    }
+
+    // test driver for each queue
+    std::vector<std::unique_ptr<scheduler_tester>> drivers;
+    drivers.reserve(paths.size());
+    for (const auto& path : paths) {
+        storage()->create(path).get();
+        drivers.push_back(std::make_unique<scheduler_tester>(
+          scheduler(), storage(), path, file_size(), num_io_ops()));
+    }
+
+    // start the driver for each io queue
+    std::vector<seastar::future<>> runners;
+    runners.reserve(drivers.size());
+    for (auto& driver : drivers) {
+        runners.push_back(driver->run());
+    }
+
+    seastar::when_all_succeed(runners.begin(), runners.end()).get();
+
+    for (auto& driver : drivers) {
+        driver->stop();
+    }
+
+    // verify
+    for (auto& driver : drivers) {
+        const auto& gen = driver->generator;
+        ASSERT_GE(gen.reads().size(), num_io_ops());
+        ASSERT_FALSE(gen.writes().empty());
+        for (const auto& read : gen.reads()) {
+            EXPECT_TRUE(gen.writes().contains(read->offset()));
+            auto* write = gen.writes().at(read->offset());
+            EXPECT_EQ(read->data(), write->data());
+        }
+    }
+
+    for (auto& driver : drivers) {
+        try {
+            seastar::remove_file(driver->queue.path().string()).get();
+        } catch (...) {
+        }
+    }
```

Comment:
Maybe stick this in a deferred block so it runs even if the test fails?

Reply:
fixed!